ampd/src/event_sub/event_stream.rs

```diff
@@ -0,0 +1,272 @@
+    tm_client: &T,
+    block_height: Result<block::Height>,
+    skip_block_fn: impl Fn() -> bool,
+) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    match (skip_block_fn(), block_height) {
+        (true, _) => Ok(vec![]),
+        (_, Ok(block_height)) => events(tm_client, block_height).await,
+        (_, Err(err)) => Err(err),
+    }
+}
+
+async fn block_stream<'a, T>(
+    tm_client: &'a T,
+    poll_interval: Duration,
+) -> Result<impl Stream<Item = Result<block::Height>> + 'a>
+where
+    T: TmClient,
+{
+    latest_block_height(tm_client)
+        .await
+        .map(|latest| BlockState {
+            current: latest,
+            latest,
+        })
+        .map(|block_state| block_state.stream(tm_client, interval(poll_interval)))
+        .map(Box::pin)
+}
+
+async fn events<T>(tm_client: &T, block_height: block::Height) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    let block_results =
+        tm_client
+            .block_results(block_height)
+            .await
+            .change_context(Error::BlockResultsQuery {
+                block: block_height,
+            })?;
+
+    let begin_block_events = block_results.begin_block_events.into_iter().flatten();
+    let tx_events = block_results
+        .txs_results
+        .into_iter()
+        .flatten()
+        .flat_map(|tx| tx.events);
+    let end_block_events = block_results.end_block_events.into_iter().flatten();
+
+    let events = begin_block_events
+        .chain(tx_events)
+        .chain(end_block_events)
+        .map(|event| {
+            Event::try_from(event).change_context(Error::EventDecoding {
+                block: block_height,
+            })
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(iter::once(Event::BlockBegin(block_height))
+        .chain(events)
+        .chain(iter::once(Event::BlockEnd(block_height)))
+        .collect())
+}
+
+#[derive(Copy, Clone)]
+struct BlockState {
+    current: block::Height,
+    latest: block::Height,
+}
+
+impl BlockState {
+    async fn update<T>(mut self, tm_client: &T, interval: &mut Interval) -> Result<Self>
+    where
+        T: TmClient,
+    {
+        while self.current > self.latest {
+            self.latest = interval
+                .tick()
+                .then(|_| latest_block_height(tm_client))
+                .await?;
+        }
+
+        self.current = self.current.increment();
+
+        Ok(self)
+    }
+
+    fn stream<'a, T>(
+        self,
+        tm_client: &'a T,
+        interval: Interval,
+    ) -> impl Stream<Item = Result<block::Height>> + 'a
+    where
+        T: TmClient,
+    {
+        futures::stream::unfold(
+            (self, tm_client, interval),
+            |(block_state, tm_client, mut interval)| async move {
+                let current = block_state.current;
+
+                match block_state.update(tm_client, &mut interval).await {
+                    Ok(block_state) => Some((Ok(current), (block_state, tm_client, interval))),
```

Comment:
this is going to emit a block 1 in the future

Reply:
How does it happen?